pkg/webhook/certs.go

```diff
@@ -0,0 +1,484 @@
+	// Use a new client so we are unaffected by the cache sync kill signal
+	cli, err := client.New(mgr.GetConfig(), client.Options{Scheme: mgr.GetScheme(), Mapper: mgr.GetRESTMapper()})
+	if err != nil {
+		return err
+	}
+	rotator := &certRotator{client: cli}
+	mgr.Add(rotator)
+
+	reconciler := &ReconcileVWH{
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+		ctx:    context.Background(),
+	}
+	if err := addController(mgr, reconciler); err != nil {
+		return err
+	}
+	return nil
+}
+
+type certRotator struct {
+	client client.Client
+}
+
+func (cr *certRotator) Start(stop <-chan (struct{})) error {
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if restart, err := cr.refreshCertIfNeeded(); err != nil {
+		crLog.Error(err, "could not refresh cert on startup")
+		return errors.Wrap(err, "could not refresh cert on startup")
+	} else if restart {
+		crLog.Info("certs refreshed, restarting server")
+		return nil
+	}
+	ticker := time.NewTicker(rotationCheckFrequency)
+
+	tickerLoop:
+	for {
+		select {
+		case <-ticker.C:
+			if restart, err := cr.refreshCertIfNeeded(); err != nil {
+				crLog.Error(err, "error rotating certs")
+			} else if restart {
+				crLog.Info("certs refreshed, restarting server")
+				break tickerLoop
+			}
+		case <-stop:
+			break tickerLoop
+		}
+	}
+
+	ticker.Stop()
+	return nil
+}
+
+// refreshCertIfNeeded returns whether the cert was refreshed and any errors
+func (cr *certRotator) refreshCertIfNeeded() (bool, error) {
+	secret := &corev1.Secret{}
+	if err := cr.client.Get(context.Background(), secretKey, secret); err != nil {
+		return false, errors.Wrap(err, "acquiring secret to update certificates")
+	}
+	if secret.Data == nil || !validCACert(secret.Data[caCertName], secret.Data[caKeyName]) {
+		crLog.Info("refreshing CA and server certs")
+		return cr.refreshCerts(true, secret)
+	}
+	// make sure our reconciler is initialized on startup (either this or the above refreshCerts() will call this)
+	if !validServerCert(secret.Data[caCertName], secret.Data[certName], secret.Data[keyName]) {
+		crLog.Info("refreshing server certs")
+		return cr.refreshCerts(false, secret)
+	}
+	crLog.Info("no cert refresh needed")
+	return false, nil
+}
+
+func (cr *certRotator) refreshCerts(refreshCA bool, secret *corev1.Secret) (bool, error) {
+	var caArtifacts *KeyPairArtifacts
+	if refreshCA {
+		var err error
+		caArtifacts, err = createCACert()
+		if err != nil {
+			return false, err
+		}
+	} else {
+		var err error
+		caArtifacts, err = buildArtifactsFromSecret(secret)
+		if err != nil {
+			return false, err
+		}
+	}
+	cert, key, err := createCertPEM(caArtifacts)
+	if err != nil {
+		return false, err
+	}
+	writeFn := func() (bool, error) {
+		if err := cr.writeSecret(cert, key, caArtifacts, secret); err != nil {
```

Comment:
maybe i missed this somewhere, but after `refreshCerts` is done, don't we need to call `injectCertToWebhook` to update the data in the vwh?

Reply:
oh wait i see it's part of the secret's reconciler

Reply:
That's because the reconciler run is triggered by both changes to the secret and to the vwh config. Therefore writing the secret will trigger an update, or (if the reconciler is not yet running) the vwh will be updated once the reconciler finishes bootstrapping.

Reply:
haha simultaneous comments, closing.